### DISCUSSION TITLE
Compiling external CPP modules

### DISCUSSION THREAD

**Discussion by steoj** [micropython/micropython#14172] [ESP8266]
Have been reading the docs on this and it seems there are two paths on buliding external C/C++ micropython modules: Make and CMake.
Trying to compile a c++ module for a port/esp8266 it seems the module is not included unless using
`make USER_C_MODULES=../../examples/usercmodule`
But then I get the error message:
```
In file included from lib/micropython/examples/usercmodule/cppexample/example.cpp:2:0:
lib/micropython/examples/usercmodule/cppexample/examplemodule.h:2:24: fatal error: py/runtime.h: No such file or directory
 #include "py/runtime.h"
```
Am I missing something, or is it just not possible to compile C++ modules as external modules for esp8266 port?

Using `USER_C_MODULES=../../examples/usercmodule/micropython.cmake` does not include any of the example modules at all.

**Comment by jimmo** [comment]
>  there are two paths on buliding external C/C++ micropython modules: Make and CMake.

@steoj Make or CMake is entirely dependent on which microcontroller you're using (because we match the vendor's SDK). For ESP8266 you must use Make.

> is it just not possible to compile C++ modules as external modules for esp8266 port?

Yeah this looks like when C++ support was added, the ESP8266 port was missed, and the C++ compilation is not inheriting the right flags (which is why it's not getting the correct include paths). You need to add something like this to `ports/esp8266/Makefile`:

```
CXXFLAGS += $(filter-out -Wmissing-prototypes -Wold-style-definition -std=gnu99,$(CFLAGS))
```

Put it just above the line:

```
# Options for mpy-cross
MPY_CROSS_FLAGS += -march=xtensa
```

**Reply by jimmo** [reply]
Raised https://github.com/micropython/micropython/pull/14178 to fix.

**Reply by steoj** [reply]
Thank you. Got a little further:
```
LINK  build/firmware.elf
xtensa-lx106-elf-ld: cannot find -lstdc++
make[1]: *** [Makefile:210: build/firmware.elf] Error 1
make[1]: Leaving directory 'lib/micropython/ports/esp8266'
make: *** [Makefile:38: firmware] Error 2
```
Seems like the `lib/micropython/examples/usercmodule/cppexample/micropython.mk` contains the line:
```
LDFLAGS_USERMOD += -lstdc++
```
Removing this seems to fix the problem and I can load, run and test the cpp function:
```
>>> import cppexample
>>> cppexample.cppfunc(2,3)
(5, 'hellocpp')
>>> 
```

Context: I have made a Makefile at the root of the project I'm working on and this makefile includes the esp8266 port make as well. 
```
USER_C_MODULES ?= $(abspath lib/micropython/examples/usercmodule)
```
is defined in this makefile

**Reply by jimmo** [reply]
@steoj How are you setting up the ESP8266 toolchain? I didn't run into this issue.

FWIW, I use the process described in the "Building with a local toolchain" section of ports/esp8266/README.md.

**Reply by steoj** [reply]
Using the same instructions but running the docker image:
```
docker run --rm -v $HOME:$HOME -u $UID -w $PWD larsks/esp-open-sdk make v=1  -j BOARD=ESP8266_GENERIC
```
Im on a Mack M3 Pro, and installing the Espressif toolchain proves to be a little difficult.  [Standard Setup of Toolchain for Mac OS](https://docs.espressif.com/projects/esp8266-rtos-sdk/en/latest/get-started/macos-setup.html) kinda works, but the compiler is throwing 
```
xtensa-lx106-elf-gcc: error: unrecognized command line option '-mforce-l32'; did you mean '--force-link'?
-e See https://github.com/micropython/micropython/wiki/Build-Troubleshooting
```
at me. The troubleshooting does not give me much.

**Reply by dpgeorge** [reply]
You need a custom version of the `xtensa-lx106-elf-gcc` compiler, which supports the command `-mforce=l32`.  

I'm not sure exactly how to get this all working on a Mac, using a Docker image is probably the easiest way to do it.  In that case you can try installing the custom compiler using the commands from `tools/ci.sh`:
```
function ci_esp8266_setup {
    sudo pip install pyserial esptool==3.3.1
    wget https://github.com/jepler/esp-open-sdk/releases/download/2018-06-10/xtensa-lx106-elf-standalone.tar.gz
    zcat xtensa-lx106-elf-standalone.tar.gz | tar x
    # Remove this esptool.py so pip version is used instead
    rm xtensa-lx106-elf/bin/esptool.py
}
```
